src/python/pants/backend/python/rules/pex.py

```diff
@@ -15,18 +15,65 @@
   DirectoriesToMerge,
   DirectoryWithPrefixToAdd,
 )
-from pants.engine.isolated_process import ExecuteProcessResult, MultiPlatformExecuteProcessRequest
+from pants.engine.isolated_process import (
+  ExecuteProcessRequest,
+  ExecuteProcessResult,
+  MultiPlatformExecuteProcessRequest,
+)
+from pants.engine.legacy.structs import PythonTargetAdaptor, TargetAdaptor
 from pants.engine.platform import Platform, PlatformConstraint
-from pants.engine.rules import optionable_rule, rule
+from pants.engine.rules import RootRule, optionable_rule, rule
 from pants.engine.selectors import Get
+from pants.util.strutil import create_path_env_var
+
+
+@dataclass(frozen=True)
+class PexRequirements:
+  requirements: Tuple[str] = ()
+
+  @classmethod
+  def create_from_adaptors(cls, adaptors: Tuple[TargetAdaptor], additional_requirements: Tuple[str]) -> 'PexRequirements':
+    all_target_requirements = []
+    for maybe_python_req_lib in adaptors:
+      # This is a python_requirement()-like target.
+      if hasattr(maybe_python_req_lib, 'requirement'):
+        all_target_requirements.append(str(maybe_python_req_lib.requirement))
+      # This is a python_requirement_library()-like target.
+      if hasattr(maybe_python_req_lib, 'requirements'):
+        for py_req in maybe_python_req_lib.requirements:
+          all_target_requirements.append(str(py_req.requirement))
+
+      all_target_requirements.extend(additional_requirements)
+    return PexRequirements(requirements=tuple(sorted(all_target_requirements)))
+
+@dataclass(frozen=True)
+class PexInterpreterContraints:
+  constraint_set: FrozenSet[str] = frozenset()
+
+  def generate_pex_arg_list(self) -> List[str]:
+    args = []
+    for constraint in self.constraint_set:
+      args.extend(["--interpreter-constraint", constraint])
+    return args
+
+  @classmethod
+  def create_from_adaptors(cls, adaptors: Tuple[PythonTargetAdaptor], python_setup: PythonSetup) -> 'PexInterpreterContraints':
+    interpreter_constraints = frozenset(
+      [constraint
+      for target_adaptor in adaptors
+      for constraint in python_setup.compatibility_or_constraints(
+        getattr(target_adaptor, 'compatibility', None)
+      )]
```

Comment:
No need for wrapping this in a list comprehension. You can use a generator expression instead, which not only means less typing but also means that the comprehension gets evaluated lazily:

```python
frozenset(x for x in my_collection)
```